rfcs/20210115-tfx-periodic-training.md

```diff
@@ -0,0 +1,445 @@
+for correct ML metadata tracking.
+
+##### Partial pipeline execution
+
+In certain backfill cases users may not want to run the entire pipeline. For
+example, when there are multiple spans in the rolling window that need to be
+backfilled, it’s not necessary to retrain the model multiple times. In such
+cases, user might want to (conceptually) divide the pipeline in two for a faster
+backfill:
+
+*   The first portion contains components like ExampleGen, StatisticsGen and
+    ExampleValidator. This part of the pipeline is scheduled for each Span.
+*   The second portion contains a Resolver plus components like Transform,
+    Trainer, Evaluator and Pusher. This part of the pipeline only needs to run
+    once for updating the model after all Spans are updated.
+
+We plan to provide the ability to specify a partial pipeline to execute at
+runtime (see Future Work below) to enable this workflow.
+
+An alternative approach is to define the two pipeline portions as TFX pipelines
+in their own right, to make backfill easier. This is not recommended because it
+makes the normal case of periodic training more brittle: the two pipelines must
+now be triggered in sequence by some non-TFX mechanism, thereby multiplying the
+possible fault conditions.
+
+##### Training Considerations for Backfill Runs
+
+When there are issues with input data which require new versions of a span, the
+models that have been trained with the problematic data may be of low quality.
+Warmstarting can propagate this damage forward in an unpredictable way. To
+recover a pipeline that uses warmstarting, it’s advisable to retrain from
+scratch. If full retraining is too costly, consider manually picking up a valid
+previous model to warmstart using
+[Importer](https://github.com/tensorflow/tfx/blob/r0.26.0/tfx/components/common_nodes/importer_node.py).
+
+#### ExampleGen Data Selection
+
+As mentioned above, each ExampleGen execution generates only one Span at a time.
+ExampleGen can be configured to process a specific Span (always taking its
+latest Version); if a Span is not specified, ExampleGen selects the input data
+corresponding to the latest Version of the latest Span by default. The latest
+here means the one that has the highest Span id.
+
+Relying on “latest Span” semantics in periodic automation introduces a new
+failure mode: if the data isn’t available as expected the pipeline will
+reprocess old data. While it’s possible to guard against this with upstream
+monitoring and alerting, we recommend specifying the span as runtime parameter
+when triggering the run. For example, for a daily cron-triggered pipeline, pass
+the scheduled date to the job for calculating target span of ExampleGen. The
+pipeline will fail when the specified span’s input data is not available. This
+provides deterministic behavior for each periodic run and makes it easier for
+the pipeline owner to identify which span needs to be backfilled after failure.
```

Comment:
Also, related to this. Are there plans to add a span based rolling resolver? Given currently there's only [this](https://github.com/tensorflow/tfx/blob/master/tfx/dsl/experimental/latest_artifacts_resolver.py) and it's based on latest available instead.

Reply:
Yep, WIP, will add a resolver that takes range_config